[CI] StandardVersusLogsIndexModeRandomDataChallengeRestIT testHistogramAggregation failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #3803 / windows-2019_checkpart1_platform-support-windows](https://gradle-enterprise.elastic.co/s/mzqc5a5nr22d2)
- [elasticsearch-periodic-platform-support #3794 / almalinux-8-aarch64_checkpart1_platform-support-arm](https://gradle-enterprise.elastic.co/s/da2w3zx3raza4)
- [elasticsearch-periodic-platform-support #3794 / debian-11_platform-support-unix](https://gradle-enterprise.elastic.co/s/5a5ft2saltwgk)
- [elasticsearch-periodic-platform-support #3791 / windows-2016_checkpart1_platform-support-windows](https://gradle-enterprise.elastic.co/s/myixvfw552tia)
- [elasticsearch-periodic #3818 / encryption-at-rest](https://gradle-enterprise.elastic.co/s/pm3rfr4qyc43q)

**Reproduction Line:**
```
gradlew ":modules:data-streams:javaRestTest" --tests "org.elasticsearch.datastreams.logsdb.qa.StandardVersusLogsIndexModeRandomDataChallengeRestIT.testHistogramAggregation" -Dtests.seed=9A0BF45D088C5D9D -Dtests.locale=ms -Dtests.timezone=US/Central -Druntime.java=22
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testHistogramAggregation'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.datastreams.logsdb.qa.StandardVersusLogsIndexModeRandomDataChallengeRestIT'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure Message:**
```
org.elasticsearch.client.ResponseException: method [PUT], host [http://[::1]:63655], URI [/_index_template/baseline-template], status line [HTTP/1.1 400 Bad Request]
{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"composable template [baseline-template] template after composition is invalid"}],"type":"illegal_argument_exception","reason":"composable template [baseline-template] template after composition is invalid","caused_by":{"type":"illegal_argument_exception","reason":"invalid composite mappings for [baseline-template]","caused_by":{"type":"illegal_argument_exception","reason":"Limit of nested fields [50] has been exceeded"}}},"status":400}
```

**Issue Reasons:**
- [main] 5 failures in test testHistogramAggregation (2.3% fail rate in 214 executions)
- [main] 3 failures in pipeline elasticsearch-periodic-platform-support (60.0% fail rate in 5 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

Failures seem to be caused by assert that tripped:

```
[2024-08-19T21:28:59,250][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [test-cluster-0] fatal error in thread [elasticsearch[test-cluster-0][search][T#28]], exiting java.lang.AssertionError: index=-1 numBits=8743	
	at org.apache.lucene.core@9.11.1/org.apache.lucene.util.FixedBitSet.prevSetBit(FixedBitSet.java:296)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.NestedObjectMapper$NestedSyntheticFieldLoader.collectChildren(NestedObjectMapper.java:444)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.NestedObjectMapper$NestedSyntheticFieldLoader.lambda$docValuesLoader$0(NestedObjectMapper.java:434)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader$ObjectDocValuesLoader.advanceToDoc(ObjectMapper.java:819)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.write(SourceLoader.java:210)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.SourceLoader$Synthetic$LeafWithMetrics.write(SourceLoader.java:158)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.NestedObjectMapper$NestedSyntheticFieldLoader.write(NestedObjectMapper.java:468)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader.write(ObjectMapper.java:871)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.write(SourceLoader.java:214)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.source(SourceLoader.java:181)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.SourceLoader$Synthetic$LeafWithMetrics.source(SourceLoader.java:146)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.prepareNonNestedHitContext(FetchPhase.java:289)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.prepareHitContext(FetchPhase.java:232)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase$1.nextDoc(FetchPhase.java:161)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhaseDocsIterator.iterate(FetchPhaseDocsIterator.java:71)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.buildSearchHits(FetchPhase.java:190)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:80)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:756)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:701)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:559)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:78)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:75)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.action.ActionRunnable$4.doRun(ActionRunnable.java:100)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:991)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)	
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)	
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)	
	at java.base/java.lang.Thread.run(Thread.java:1570)	
	Suppressed: java.lang.IllegalStateException: Failed to close the XContentBuilder	
		at org.elasticsearch.xcontent@8.16.0-SNAPSHOT/org.elasticsearch.xcontent.XContentBuilder.close(XContentBuilder.java:1281)	
		at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.source(SourceLoader.java:180)	
		... 20 more	
	Caused by: java.io.IOException: Unclosed object or array found	
		at org.elasticsearch.xcontent.impl@8.16.0-SNAPSHOT/org.elasticsearch.xcontent.provider.json.JsonXContentGenerator.close(JsonXContentGenerator.java:599)	
		at org.elasticsearch.xcontent@8.16.0-SNAPSHOT/org.elasticsearch.xcontent.XContentBuilder.close(XContentBuilder.java:1279)	
		... 21 more
```

--- a/server/src/main/java/org/elasticsearch/index/mapper/NestedObjectMapper.java
+++ b/server/src/main/java/org/elasticsearch/index/mapper/NestedObjectMapper.java
@@ -441,7 +441,7 @@ public class NestedObjectMapper extends ObjectMapper {
         }
 
         private List<Integer> collectChildren(int parentDoc, BitSet parentDocs, DocIdSetIterator childIt) throws IOException {
-            assert parentDocs.get(parentDoc) : "wrong context, doc " + parentDoc + " is not a parent of " + nestedTypePath;
+            assert parentDoc < 0 || parentDocs.get(parentDoc) : "wrong context, doc " + parentDoc + " is not a parent of " + nestedTypePath;
             final int prevParentDoc = parentDoc > 0 ? parentDocs.prevSetBit(parentDoc - 1) : -1;
             int childDocId = childIt.docID();
             if (childDocId <= prevParentDoc) {

